i18n: Provision new `number-formatters` package

chriskmnds opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!-- Would you like this feature to be tested by Beta testers?
Please add testing instructions to projects/plugins/jetpack/to-test.md in a new commit as part of your PR. -->
<!-- a12s: If you have an expected version that you're aiming for the PR to add, please use the Milestone field to communicate it. If you leave it blank, that indicates there isn't a preference. -->

See PT: pxLjZ-9ME-p2
Part of addressing https://github.com/Automattic/i18n-issues/issues/942

## Proposed changes:
<!--- Explain what functional changes your PR includes -->
Introduces a new js-package `number-formatters` for formatting numbers/currencies/percentages/etc. in a unified way across wpcom frontends.

This is a skeleton PR with base configs to provision the new package.

### Other information:

- [x] Have you written new tests for your changes, if applicable?
- [x] Have you checked the E2E test CI results, and verified that your changes do not break them?
- [x] Have you tested your changes on WordPress.com, if applicable (if so, you'll see a generated comment below with a script to run)?

## Jetpack product discussion
<!-- If you're an Automattician, include a shortlink to the p2 discussion with Jetpack Product here. -->
<!-- Make sure any changes to existing products have been discussed and agreed upon -->

## Does this pull request change what data or activity we track or use?
<!--- If so, please add the "[Status] Needs Privacy Updates" label and explain what changes there are. -->
<!--- Check existing Jetpack support documents for a preview of the information we need. -->

## Testing instructions:
<!-- If you were reviewing this PR, how would you like the instructions to be presented? -->
<!-- Please include detailed testing steps, explaining how to test your change. -->
<!-- Bear in mind that context you working on is not obvious for everyone.  -->
<!-- Adding "simple" configuration steps will help reviewers to get to your PR as quickly as possible. -->
<!-- "Before / After" screenshots can also be very helpful when the change is visual. -->

* Build passing
* Confirm package release on NPM once shipped

chriskmnds committed:
i18n: Provision new number-formatters package

chriskmnds committed:
install

anomiex commented on projects/js-packages/number-formatters/eslint.config.mjs:
```diff
@@ -0,0 +1,27 @@
+export default [
+	...makeBaseConfig( import.meta.url ),
+	{
+		languageOptions: {
+			parserOptions: {
+				project: './tsconfig.json',
+			},
+		},
```
Is this needed? I don't see it in any of our other TypeScript-using projects. The only one that has something like this is `projects/plugins/boost/eslint.config.mjs` because it has an extra `./tsconfig.eslint.json`.

P.S. You're welcome to do a `tsconfig.eslint.json` like Boost does if that's appropriate. It might be needed if you want tests linted properly?

anomiex commented on projects/js-packages/number-formatters/eslint.config.mjs:
```diff
@@ -0,0 +1,27 @@
+			},
+		},
+		files: [ '**/*.ts', '**/*.tsx' ],
+		plugins: {
+			'@typescript-eslint': true,
+			jsdoc: true,
+		},
```
This won't work with eslint 9, and I think it's redundant to what our base config already provides.

anomiex commented on projects/js-packages/number-formatters/eslint.config.mjs:
```diff
@@ -0,0 +1,27 @@
+			jsdoc: true,
+		},
+		rules: {
+			'jsdoc/require-jsdoc': 'off',
+			'jsdoc/require-description': 'off',
+			'jsdoc/require-returns': 'off',
+			'jsdoc/require-returns-check': 'off',
+			'jsdoc/require-param': 'off',
+			'jsdoc/require-param-description': 'off',
+			'jsdoc/require-param-type': 'off',
```
Personally I'd recommend you not disable all these, as good documentation will help reusers know how to use things and these set the low bar of "documentation exists" for that. But I won't try to force it if you really don't want these.

P.S. Our base config already turns off `jsdoc/require-param-type` for TypeScript files (as well as `jsdoc/require-property-type` and `jsdoc/require-returns-type`), since TypeScript already provides type declarations.

chriskmnds replied:
I think this was just the result of battling it out to remove `jsdoc/require-returns`. We don't require params and return definitions in the JS docs in Calypso, as those are rather accurately defined from TS types for the most part. In cases where ambiguity might exist e.g. a string return that's the transformation of something, I'd normally define these, but in general, I'd just rely on TS.

Happy to remove these if it's a convention in JP.

anomiex replied:
For TypeScript code, our standard configuration shouldn't require the type be specified in the jsdoc but will still want a human-readable description of the parameters and return value. See files in `projects/js-packages/publicize-components/` for example.

chriskmnds replied:
Fair. Removed (upcoming changes)

anomiex commented on projects/js-packages/number-formatters/jest.config.cjs:
```diff
@@ -0,0 +1,9 @@
+
+module.exports = {
+	...baseConfig,
+	collectCoverageFrom: [
+		'<rootDir>/src/**/*.{js,mjs,cjs,jsx,ts,tsx,mts,cts}',
+		...baseConfig.collectCoverageFrom,
+	],
```
I believe this `collectCoverageFrom` is redundant, `baseConfig.collectCoverageFrom` already contains `'<rootDir>/src/**/*.{js,mjs,cjs,jsx,ts,tsx,mts,cts}'`. 🙂

chriskmnds replied:
Some of my comments may not have come through, or may come later 😬 

I asked a question from the editor, whether this file is needed at all.

anomiex replied:
I think some config file is needed to load the basic configuration, but many are as simple as
```
const baseConfig = require( 'jetpack-js-tools/jest/config.base.js' );

module.exports = {
        ...baseConfig,
};
```

chriskmnds replied:
got it. thanks

anomiex commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,38 @@
+{
+	"name": "@automattic/jetpack-number-formatters",
```
You're welcome to rename this to something like `@automattic/number-formatters` if you want. Including "jetpack-" isn't required. 🙂

anomiex commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,38 @@
+		"@babel/runtime": "^7"
+	},
+	"devDependencies": {
+		"@babel/core": "7.26.0",
```
The rest of the monorepo is on 7.26.10.
```suggestion
		"@babel/core": "7.26.10",
```

anomiex commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,38 @@
+	},
+	"type": "module",
+	"exports": {
+		".": "./index.ts"
```
This is one of the things I was referring to in https://github.com/Automattic/jetpack/pull/42317#issuecomment-2730139006 about setting TypeScript up correctly. The package should offer JS modules and `.d.ts` types rather than the `.ts` file correctly.

To do that, you'd do like this here:
```suggestion
		".": {
			"jetpack:src": "./index.ts",
			"types": "./build/index.d.ts",
			"default": "./build/index.js"
		}
```
(and then set things up elsewhere to produce those `./build/` files)

anomiex commented on projects/js-packages/number-formatters/.gitattributes:
```diff
@@ -0,0 +1,8 @@
+# Files not needed to be distributed in the package.
+.gitattributes    export-ignore
+node_modules      export-ignore
+
```
Regarding the setting things up correctly mentioned in https://github.com/Automattic/jetpack/pull/42317#issuecomment-2730139006, you'll want to add something like this in here
```suggestion

# Files to include in the mirror repo, but excluded via gitignore
# Remember to end all directories with `/**` to properly tag every file.
/build/**         production-include

```

You'll also need to add a `projects/js-packages/number-formatters/.gitignore` file with
```
build/
```

anomiex commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,38 @@
+		".": "./index.ts"
+	},
+	"scripts": {
+		"build": "pnpm run compile-ts",
```
Continuing on the theme of "setting it up correctly" from https://github.com/Automattic/jetpack/pull/42317#issuecomment-2730416802, you'll likely want to change this something like
```suggestion
		"build": "pnpm run clean && pnpm run compile-ts",
		"clean": "rm -rf build/",
```

anomiex commented on projects/js-packages/number-formatters/tsconfig.json:
```diff
@@ -0,0 +1,4 @@
+{
+	"extends": "jetpack-js-tools/tsconfig.base.json",
+	"include": [ "./index.ts", "./src" ]
+}
```
And finally for the setting things up correctly, you'll need to have this look something like
```suggestion
{
	"extends": "jetpack-js-tools/tsconfig.tsc.json",
	"include": [ "./index.ts" ],
	"compilerOptions": {
		"typeRoots": [ "./node_modules/@types/" ],
		"sourceMap": false,
		"outDir": "./build/",
		"target": "es2022",
	}
}
```
You're welcome to use whichever target is appropriate, rename "build" to "dist" if you prefer that, etc.

chriskmnds replied:
Ok. Added. Indeed I changed "build" to "dist".

The only problem is changing `"extends": "jetpack-js-tools/tsconfig.base.json"` to `"extends": "jetpack-js-tools/tsconfig.tsc.json"`. It gives the error I mentioned above

chriskmnds replied:
Ok - my replies from the editor may not be coming through. I said earlier:

with the suggested change in "extends", I get the following error in `index.ts`:

> Relative import paths need explicit file extensions in ECMAScript imports when '--moduleResolution' is 'node16' or 'nodenext'. Did you mean './src/format-number.js'?

anomiex replied:
That's expected, and you should make the change it calls for.

What's going on here is that Node [requires the extension for esm imports](https://nodejs.org/docs/latest-v22.x/api/esm.html#mandatory-file-extensions), while bundlers like webpack typically don't.

`tsc` doesn't try to rewrite the imports at all, it just outputs what its given. If the tsconfig has moduleResolution set to "nodenext" rather than "bundler", it issues that error to flag that the import won't work if the output is going to be run by Node. When you're writing a library like this one, you don't know whether the project using your library is going to be using "nodenext" or "bundler", so [TypeScript recommends using the stricter setting for best compatibility](https://www.typescriptlang.org/docs/handbook/modules/theory.html#module-resolution-for-libraries).

chriskmnds replied:
That's interesting. Thanks for explaining. I did come across this in the long distant past. I need to get my head around it.

I don't exactly understand this part:

> When you're writing a library like this one, you don't know whether the project using your library is going to be using "nodenext" or "bundler"

How do I not know that? Does that make every package we export from wp-calypso faulty in that sense (cause this is the first time I come across this)?

Sorry for my ignorance. I'll try and sum it up in my head

anomiex replied:
> Does that make every package we export from wp-calypso faulty in that sense (cause this is the first time I come across this)?

Maybe? Or maybe I've misunderstood something at some point. But most likely it just hasn't come up because either all reusers are processing the packages via a bundler like webpack or they're using `tsc` with `moduleResolution` set to "bundler" (and aren't running the package via node directly either).

chriskmnds replied:
Interesting. That may well be the case. Thanks for clarifying and sharing the links above.

anomiex reviewed (commented):
In order for the monorepo tooling to work, you'll need to add a few additional files.

* A `composer.json` with some configuration settings.
  <details>
   
  This should do it:
 
  ```json
  {
      "name": "automattic/js-number-formatter",
      "description": "Number formatting utilities.",
      "license": "GPL-2.0-or-later",
      "require": {},
      "require-dev": {
          "automattic/jetpack-changelogger": "@dev"
      },
      "scripts": {
          "build-production": [
              "pnpm run build"
          ],
          "build-development": [
              "pnpm run build"
          ],
          "test-js": [
              "pnpm run test"
          ],
          "test-coverage": [
              "pnpm run test-coverage"
          ]
      },
      "repositories": [
          {
              "type": "path",
              "url": "../../packages/*",
              "options": {
                  "monorepo": true
              }
          }
      ],
      "extra": {
          "autotagger": true,
          "npmjs-autopublish": true,
          "mirror-repo": "Automattic/number-formatter",
          "textdomain": "number-formatter",
          "changelogger": {
              "link-template": "https://github.com/Automattic/number-formatter/compare/${old}...${new}"
          }
      }
  }
  ````
  
  The `name` field here doesn't matter. The really important stuff is in `scripts` and `extra`.
  
  Of particular note in there is the reference to a mirror repo, which as specified would be https://github.com/Automattic/number-formatter. If you'd rather have a different mirror repo name (under the Automattic organization), that's fine. Once the name is settled, I can create it for you or you can try following the instructions in https://github.com/Automattic/jetpack/blob/trunk/docs/monorepo.md#mirror-repositories. 🙂
  
  </details>
* An empty file at `changelog/.gitkeep`.
* The `.gitignore` file mentioned in one of the inline comments.

chriskmnds committed:
address feedback

chriskmnds commented on projects/js-packages/number-formatters/composer.json:
```diff
@@ -0,0 +1,41 @@
+	"extra": {
+		"autotagger": true,
+		"npmjs-autopublish": true,
+		"mirror-repo": "Automattic/number-formatters",
```
@anomiex  Is this sound? You suggested `Automattic/number-formatter` (singular) in your code, which makes me wonder if that would be a better name overall. I've used "number-formatters" plural throughout 🤔

anomiex replied:
Either is fine as far as I'm concerned. 🙂

anomiex replied:
Plural to match the npm package name is probably the best way to go. :+1:

chriskmnds committed:
changelog

chriskmnds commented:
@anomiex Thanks for the help/feedback! :-)

I've addressed most comments and left a couple of questions above. I'd be happy to get a bare-bones package shipped (with no more functionality), just to have a base-line setup to refer to. I'd like that if acceptable. Bringing more code into this PR may require additional dependencies, etc.

anomiex commented on pnpm-lock.yaml:
Something seems to be going wrong with this file, it's getting massively reformatted.

Your best bet may be to `git checkout origin/trunk pnpm-lock.yaml` and then `pnpm dedupe` to regenerate it.

chriskmnds replied:
👍

anomiex commented on projects/js-packages/number-formatters/CHANGELOG.md:
```diff
@@ -0,0 +1,12 @@
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+
+## [0.1.0] - 2024-03-18
```
Since there's no link for the 0.1.0 version in the file, we need to delink it here.
```suggestion
## 0.1.0 - 2024-03-18
```

anomiex commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,43 @@
+{
+	"name": "@automattic/number-formatters",
+	"version": "0.1.0",
```
The "changelogger validity" warning from CI here should be fixed by the suggestion I made for CHANGELOG.md.

chriskmnds committed:
Update projects/js-packages/number-formatters/CHANGELOG.md

Co-authored-by: Brad Jorsch <anomiex@users.noreply.github.com>

chriskmnds committed:
address feedback

chriskmnds committed:
regenerate pnpm-lock

chriskmnds committed:
update js docs

chriskmnds commented:
@anomiex you are trully awesome. Thanks for the continued help and feedback :-)

chriskmnds commented:
I think this looks good to go now, thanks to @anomiex. 
I'd like to see the package on NPM next so I can start practically using it

I created the mirror repo [here](https://github.com/Automattic/number-formatters), although I couldn't figure out steps 4 and 6 from the instruactions:

> iv. Make sure that [matticbot](https://github.com/matticbot) can push to the repo. You would do this here: https://github.com/Automattic/example-repository-name/settings/branches - creating a new branch protection rule where only Matticbot (and whoever needs access to push, for example Garage) can push to that repository.

I created the ruleset, but that's as far as I got. I couldn't find the Automattic Bot app to add to the bypass list (if that's what I'm supposed to do)

> vi. Create any secrets needed (e.g. for Autotagger or Npmjs-Autopublisher). See PCYsg-xsv-p2#mirror-repo-secrets for details.

anomiex commented:
The step about matticbot usually doesn't actually need you to do anything.

I'll go ahead and add the secrets.

anomiex reviewed (commented):
You still need to add a `.gitignore` file with contents like
```
/dist/
```

chriskmnds committed:
.gitignore

chriskmnds commented:
Thanks @anomiex. Up to you how best to proceed from here. We can flesh this out into the full package or ship the skeleton and add the logic next. Come to think of it, it doesn't make a lot of difference on my end :-)

> You still need to add a `.gitignore` file with contents like
👍

anomiex reviewed (approved):
Looks good to me to merge and iterate.

If you want a version actually published to npm after merging this, you'll want to follow the instructions in PCYsg-Np7-p2 to have a release tagged. Feel free to ask in Slack (#jetpack-releases) if you want help with that.

chriskmnds commented:
I get these E2E tests consistently failing, although they don't look related. Do I proceed with merging?

chriskmnds commented:
Looks like the "Push to mirror repos" job failing https://github.com/Automattic/jetpack/actions/runs/14065109246/job/39386553959

anomiex commented:
Yeah, "required" is the real blocker. Unfortunately some of these E2Es are a bit flaky and no one has stepped up to fix them.

> Looks like the "Push to mirror repos" job failing https://github.com/Automattic/jetpack/actions/runs/14065109246/job/39386553959

Whatever you did for the "Make sure that [matticbot](https://github.com/matticbot) can push to the repo" step somehow prevented it. Really, for repos under the Automattic organization, it just works by default anyway. I made #42702 to update that doc, and fixed the mirror repo settings so the re-run at https://github.com/Automattic/jetpack/actions/runs/14065109246/job/39388963330 worked. :+1:

anomiex commented on projects/js-packages/number-formatters/README.md:
````diff
@@ -0,0 +1,11 @@
+## Usage
+
+```typescript
+import { formatNumber } from '@automattic/jetpack-number-formatters';
````
Oops, we missed this earlier. :sweat_smile: Something for a followup.
```suggestion
import { formatNumber } from '@automattic/number-formatters';
```

chriskmnds replied:
Gotcha. We'll have a PR soon with lots of changes here :-)

manzoorwanijk commented on projects/js-packages/number-formatters/package.json:
```diff
@@ -0,0 +1,43 @@
+	"type": "module",
+	"exports": {
+		".": {
+			"jetpack:src": "./index.ts",
+			"types": "./dist/index.d.ts",
+			"default": "./dist/index.js"
+		}
+	},
+	"scripts": {
+		"build": "pnpm run clean && pnpm run compile-ts",
+		"clean": "rm -rf dist/",
+		"compile-ts": "tsc --pretty",
```
Do we really need a build step here?

Is the package used outside Jetpack? If not, then may be we can only have `"jetpack:src"` and remove everything else?

chriskmnds replied:
Hi!

> Is the package used outside Jetpack?

Yes, intended for use outside.

anomiex replied:
Personally, I'd say if it has TS and we're publishing it to npm then it probably should have a build script set up. If we really don't want it to be possibly used externally, we probably shouldn't be publishing it in the first place.

manzoorwanijk replied:
Yes, I think the same.
